MaxText/convert_llama_ckpt.py

```diff
@@ -0,0 +1,340 @@
+    wv = np.reshape(wv, [base_num_query_heads * head_dim, base_num_kv_heads, head_dim])
+
+    w_post = np.concatenate(
+        [
+            var[f'layers.{layer_idx}.attention.wo.weight'].type(torch.float16).numpy()
+            for var in pytorch_vars
+        ],
+        axis=1,
+    )
+
+    w_post = np.reshape(w_post, [base_num_query_heads * head_dim, base_num_query_heads, head_dim])
+
+
+    self_attention['query']['kernel'].append(wq)
+    self_attention['key']['kernel'].append(wk)
+    self_attention['value']['kernel'].append(wv)
+    self_attention['out']['kernel'].append(w_post)
+
+    wi_0 = np.concatenate([var[f'layers.{layer_idx}.feed_forward.w1.weight'].type(torch.float16).numpy()
+                           for var in pytorch_vars], axis=0).transpose()
+    wi_1 = np.concatenate([var[f'layers.{layer_idx}.feed_forward.w3.weight'].type(torch.float16).numpy()
+                           for var in pytorch_vars], axis=0).transpose()
+    wo = np.concatenate([var[f'layers.{layer_idx}.feed_forward.w2.weight'].type(torch.float16).numpy()
+                                 for var in pytorch_vars], axis=1).transpose()
+    pre_self_attention_layernorm = pytorch_vars[0][f'layers.{layer_idx}.attention_norm.weight'].type(torch.float16).numpy()
+    post_self_attention_layernorm = pytorch_vars[0][f'layers.{layer_idx}.ffn_norm.weight'].type(torch.float16).numpy()
+
+
+
+    layer_weight['mlp']['wi_0']['kernel'].append(wi_0)
+    layer_weight['mlp']['wi_1']['kernel'].append(wi_1)
+    layer_weight['mlp']['wo']['kernel'].append(wo)
+    layer_weight['pre_self_attention_layer_norm']['scale'].append(pre_self_attention_layernorm)
+    layer_weight['post_self_attention_layer_norm']['scale'].append(post_self_attention_layernorm)
+
+
+  self_attention['query']['kernel'] = np.array(self_attention['query']['kernel'])
+  self_attention['key']['kernel'] = np.array(self_attention['key']['kernel'])
+  self_attention['value']['kernel'] = np.array(self_attention['value']['kernel'])
+  self_attention['out']['kernel'] = np.array(self_attention['out']['kernel'])
+  self_attention['query']['kernel'] = np.transpose(self_attention['query']['kernel'],axes=(1, 0, 2, 3))
+  self_attention['key']['kernel'] = np.transpose(self_attention['key']['kernel'],axes=(1, 0, 2, 3))
+  self_attention['value']['kernel'] = np.transpose(self_attention['value']['kernel'],axes=(1, 0, 2, 3))
+  #layers, base_num_query_heads * head_dim, base_num_query_heads, head_dim =>
+  #base_num_query_heads, layers,head_dim, base_num_query_heads * head_dim
+  self_attention['out']['kernel'] = np.transpose(self_attention['out']['kernel'],axes=(2, 0, 3, 1))
+
+  #scale the query weights
+  self_attention['query']['kernel'] = self_attention['query']['kernel']/np.sqrt(head_dim)
+
+  jax_weights['decoder']['decoder']['self_attention'] = self_attention
+
+
+
+
+
+  layer_weight['mlp']['wi_0']['kernel'] = np.array(layer_weight['mlp']['wi_0']['kernel'])
```

Comment:
Sorry my mistake was I didn't realize that these were numpys instead of jax numpy's. I think the checkpoints should be saved as jax numpys...

Reply:
if we try to do that then the checkpoint conversion will not run on a v4-8, too much memory requirement

Reply:
I commented somewhere else but I think the orbax checkpoint should just be jnp.array's instead of np.array's.

Reply:
Right now, I don't believe we will properly "load sharded"!

Reply:
converted to jax.numpy.array